python/ray/data/read_api.py

```diff
@@ -2487,31 +2487,76 @@ def from_arrow_refs(
     )
 
 
+_DATABRICKS_SPARK_DATAFRAME_CHUNK_BYTES = 32 * 1024 * 1024
+
+
 @PublicAPI
 def from_spark(
     df: "pyspark.sql.DataFrame",
     *,
-    parallelism: Optional[int] = None,
+    parallelism: int = -1,
     override_num_blocks: Optional[int] = None,
 ) -> MaterializedDataset:
     """Create a :class:`~ray.data.Dataset` from a
     `Spark DataFrame <https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.DataFrame.html>`_.
 
     Args:
-        df: A `Spark DataFrame`_, which must be created by RayDP (Spark-on-Ray).
+        df: Any `Spark DataFrame` if you run Ray application with Ray on spark,
+            or a `Spark DataFrame`_ which must be created by RayDP (Spark-on-Ray).
         parallelism: This argument is deprecated. Use ``override_num_blocks`` argument.
         override_num_blocks: Override the number of output blocks from all read tasks.
             By default, the number of output blocks is dynamically decided based on
             input data size and available resources. You shouldn't manually set this
             value in most cases.
 
     Returns:
-        A :class:`~ray.data.MaterializedDataset` holding rows read from the DataFrame.
+        If you run Ray application with Ray on spark, it returns a ray dataset, the
+        data is cached in spark distributed cache system, but is not cached
+        in ray object store, otherwise it returns
+        a :class:`~ray.data.MaterializedDataset` holding rows read from the DataFrame.
     """  # noqa: E501
-    import raydp
+    from ray.util.spark.utils import is_in_databricks_runtime
 
     parallelism = _get_num_output_blocks(parallelism, override_num_blocks)
-    return raydp.spark.spark_dataframe_to_ray_dataset(df, parallelism)
+
+    if is_in_databricks_runtime():
+        from ray.data.datasource.spark_datasource import validate_requirements, SparkDatasource
+        from ray.util.spark.databricks_hook import get_databricks_function
+
+        validate_requirements()
+
+        # Ray on spark
+        datasource = SparkDatasource(df, _DATABRICKS_SPARK_DATAFRAME_CHUNK_BYTES)
+        if parallelism == -1:
+            parallelism = datasource.num_chunks
+        dataset = read_datasource(
+            datasource=datasource,
+            override_num_blocks=parallelism,
+        )
+
+        def cleanup_hook():
+            datasource.dispose_spark_cache()
+
+        dataset._cleanup_hook = cleanup_hook
```

Comment:
This way to register the cleanup hook looks a bit hacky to me. Can we avoid doing this?

Reply:
it seems hard. I don't want to modify `read_datasource` code.

I also tried directly overwrite `dataset_obj.__del__` but it doesn't work, we have to overwrite `Dataset.__del__` then it ends up with the approach I used in this PR.

Reply:
Got it. I think we need to add a proper interface to Datasource to allow clean up resource, and we are running short of time before Ray 2.10 release. So here is my proposal:

Option 1: wait for clean up resource interface added for Datasource, then merge this PR for 2.11 release.

Option 2: Allow unpersist Spark Dataframe from Databricks side. So user can do something like:

```
spark_df = ...
ray_ds = ray.data.from_spark(spark_df)
... # run inference or doing model training on ray_ds
del ray_ds
unpersist(spark_df)
```

Reply:
We might have option 3:
Adding `__del__`  for SparkDatasource, and unpersist chunks there, I will test this approach (I think it should work)

Reply:
> Adding __del__ for SparkDatasource, and unpersist chunks there, I will test this approach (I think it should work)

Cool, this SGTM.